app/components/mappingsTable/mappingsTable.component.js

```diff
@@ -0,0 +1,224 @@
+    }
+    /*Edit logic*/
+    function startEdit(mapping, index){
+        mapping.edit = true;
+        vm.mappingsCache[index].termQuery = mapping.conceptReferenceTerm.code;
+        vm.mappingsCache[index].suggestions = [];
+        vm.incrementEditCount();
+    }
+    function confirmEdit(mapping, index){
+        if(validMapping(vm.mappingsCache[index])){
+            if(angular.isDefined(vm.mappingsCache[index].conceptReferenceTerm.uuid)){
+                executeEdit(mapping, index);
+            } else {
+                mapping.showConfirmDialog = true;
+            }
+        }
+    }
+    function cancelEdit(mapping, index){
+        mapping.edit = false;
+        vm.mappingsCache[index] = mapping;
+        vm.decrementEditCount();
+    }
+    function executeEdit(mapping, index){
+        mapping = vm.mappingsCache[index];
+        mapping.edit = false;
+        update();
+        vm.decrementEditCount();
+    }
+    /*Adding new mapping logic*/
+    function startAdd(){
+        vm.addView = true;
+        vm.incrementEditCount();
+    }
+    function confirmAdd(){
+        if(validMapping(vm.newMapping)){
+            if(angular.isDefined(vm.newMapping.conceptReferenceTerm.uuid)){
+                executeAdd();
+            } else {
+                vm.newMapping.showConfirmDialog = true;
+            }
+        }
+    }
+    function executeAdd(){
+        vm.mappings.push({conceptMapType: vm.newMapping.conceptMapType,
+                        conceptReferenceTerm: vm.newMapping.conceptReferenceTerm});
+        vm.update();
+        resetAdd();
+    }
+    function cancelAdd(){
+        resetAdd();
+    }
+    function resetAdd(){
+        vm.addView = false;
+        vm.newMapping = createNewMapping();
+        vm.decrementEditCount();
+    }
+    function createNewMapping(){
+        return {
+            conceptMapType: {},
+            conceptReferenceTerm: {
+                conceptSource: vm.sources[0]
+            },
+            termQuery: ""
+        }
+    }
+    /*adding and editing helper functions*/
+    function validMapping(map){
+        //is MapType defined? must already exist
+        if(angular.isUndefined(map.conceptMapType.uuid)){
+            openmrsNotification.error("Invalid mapping: relationship is not specified")
+            return false;
+        }
+        //fake source "search all sources..." has uuid === ""
+        if(map.conceptReferenceTerm.conceptSource.uuid === ""){
+            openmrsNotification.error("Invalid mapping: source is not specified");
+            return false;
+        }
+        //is already existing RefTerm defined?
+        var defRefTerm = angular.isDefined(map.conceptReferenceTerm.uuid);
+        //or else is new Term defined?
+        var defNewRefTerm = angular.isDefined(map.conceptReferenceTerm.conceptSource.uuid)
+                            &&angular.isDefined(map.conceptReferenceTerm.code);
+        if (defRefTerm||defNewRefTerm) return (defRefTerm||defNewRefTerm);
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Expected '{' and instead saw 'return'.](https://www.codacy.com/app/openmrs/openmrs-owa-conceptdictionary/file/2741260425/issues/source?bid=3196969&fileBranchId=3354756#l189)